src/shared/mount-util.c

```diff
@@ -861,10 +884,30 @@ static int mount_in_namespace_new(
         if (r == 0) {
                 errno_pipe_fd[0] = safe_close(errno_pipe_fd[0]);
 
-                if (make_file_or_directory)
-                        (void) mkdir_p(dest, 0755);
+                if (is_image) {
+                        if (make_file_or_directory)
+                                (void) mkdir_p(dest, 0755);
+
+                        r = dissected_image_mount(
+                                        img,
+                                        dest,
+                                        /* uid_shift= */ UID_INVALID,
+                                        /* uid_range= */ UID_INVALID,
+                                        /* userns_fd= */ -EBADF,
+                                        /* flags= */ read_only ? DISSECT_IMAGE_READ_ONLY : 0);
+                } else {
+                        if (make_file_or_directory) {
+                                (void) mkdir_parents(dest, 0755);
+                                (void) make_mount_point_inode_from_stat(chased_src_st, dest, 0700);
+                        }
 
-                if (move_mount(new_mount_fd, "", -EBADF, dest, MOVE_MOUNT_F_EMPTY_PATH) < 0) {
+                        r = RET_NERRNO(move_mount(new_mount_fd,
+                                                  "",
+                                                  -EBADF,
+                                                  dest,
+                                                  MOVE_MOUNT_F_EMPTY_PATH));
+                }
+                if (r < 0) {
```

Comment:
tbf i find the two codepaths in mount_in_namespace_new() (one for image based mounts, the other for bind mounts) sufficiently different so that they should be two functions.

maybe share the errno pipe read code below by making that another helper func (`wait_for_child_with_errno_pipe()`?), but otherwise keep the two code paths in two separate functions

Reply:
I tried but it looked more complex to me, it's just 2 alternative API calls after all, so I'd rather keep it as-is, seems clearer to me